File: app/src/main/java/de/passwordvault/view/settings/activity_licenses/LicensesViewModel.java
package de.passwordvault.view.settings.activity_licenses;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.lifecycle.ViewModel;
import java.util.ArrayList;
import de.passwordvault.R;
import de.passwordvault.view.utils.Utils;


/**
 * Class implements the view model for the activity displaying the licenses of the used software.
 *
 * @author  Christian-2003
 * @version 3.7.0
 */
public class LicensesViewModel extends ViewModel {

    /**
     * Class models a used software with a license that needs to be displayed within the app.
     */
    public static class UsedSoftware {

        /**
         * Attribute stores the name of the used software.
         */
        @NonNull
        private final String name;

        /**
         * Attribute stores the version of the used software.
         */
        @NonNull
        private final String version;

        /**
         * Attribute stores the name of the raw resource file in which the software license is
         * stored.
         */
        @NonNull
        private final String licenseResourceFile;

        /**
         * Attribute stores the name of the license used by the software.
         */
        @NonNull
        private final String licenseName;


        /**
         * Constructor instantiates a new used software instance.
         *
         * @param name                  Name of the used software.
         * @param version               Version of the used software.
         * @param licenseResourceFile   Name of the raw resource file in which the license text is
         *                              stored.
         * @param licenseName           Name of the license used by the software.
         */
        public UsedSoftware(@NonNull String name, @NonNull String version, @NonNull String licenseResourceFile, @NonNull String licenseName) {
            this.name = name;
            this.version = version;
            this.licenseResourceFile = licenseResourceFile;
            this.licenseName = licenseName;
        }


        /**
         * Method returns the name of the used software.
         *
         * @return  Name of the used software.
         */
        @NonNull
        public String getName() {
            return name;
        }

        /**
         * Method returns the version of the used software.
         *
         * @return  Version of the used software.
         */
        @NonNull
        public String getVersion() {
            return version;
        }

        /**
         * Method returns the name of the raw resource file in which the license text of the software
         * is stored.
         *
         * @return  Name of the raw resource file containing the license text.
         */
        @NonNull
        public String getLicenseResourceFile() {
            return licenseResourceFile;
        }

        /**
         * Method returns the name of the license used by the software.
         *
         * @return  Name of the license used by the software.
         */
        @NonNull
        public String getLicenseName() {
            return licenseName;
        }

    }


    /**
     * Attribute stores a list of the software used within the app.
     */
    @Nullable
    private ArrayList<UsedSoftware> usedSoftware;


    /**
     * Constructor instantiates a new view model.
     */
    public LicensesViewModel() {
        usedSoftware = null;
    }


    /**
     * Method returns a list of the used software.
     *
     * @return  List of the used software.
     */
    @NonNull
    public ArrayList<UsedSoftware> getUsedSoftware() {
        if (usedSoftware == null) {
            loadUsedSoftware();
        }
        return usedSoftware;
    }


    /**
     * Method loads the license text for the passed used software from a raw resource. The return
     * value can be {@code null} if no license text is available.
     *
     * @param software  Used software for which to load the license text.
     * @return          License text for the used software.
     */
    public String getLicenseText(UsedSoftware software) {
        return Utils.readRawResource(software.getLicenseResourceFile());
    }


    /**
     * Method loads the used software from a raw resource file. After the method finishes execution,
     * the {@link #usedSoftware} is not null.
     */
    private void loadUsedSoftware() {
        String csv = Utils.readRawResource(R.raw.licenses);
        usedSoftware = new ArrayList<>();
        if (csv == null) {
            return;
        }
        String[] software = csv.split("\n");
        for (String s : software) {
            String[] attributes = s.split(",");
            if (attributes.length == 4) {
                usedSoftware.add(new UsedSoftware(attributes[0], attributes[1], attributes[2], attributes[3]));
            }
        }
    }

}

File: app/src/main/java/de/passwordvault/view/settings/activity_licenses/LicensesViewModel.kt
package de.passwordvault.view.settings.activity_licenses

import android.app.Application
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import de.passwordvault.R
import de.passwordvault.model.licenses.License
import kotlinx.coroutines.launch
import java.io.InputStream


/**
 * Class implements the view model for the view displaying all software used by the app.
 *
 * @author  Christian-2003
 * @since   1.0.0
 */
class LicensesViewModel(application: Application): AndroidViewModel(application) {

    /**
     * Attribute stores a list of all licenses used.
     */
    var licenses: List<License> by mutableStateOf(emptyList())

    /**
     * Attribute stores whether the licenses are currently being loaded.
     */
    var isLoading: Boolean by mutableStateOf(false)

    /**
     * Attribute stores the name of the license currently displayed in a dialog.
     */
    var displayedLicenseName: String? by mutableStateOf(null)

    /**
     * Attribute stores the text of the license currently displayed in a dialog.
     */
    var displayedLicenseText: String? by mutableStateOf(null)


    /**
     * Method initializes the view model.
     */
    fun init() {

    }


    /**
     * Method loads all licenses.
     */
    fun loadLicenses() = viewModelScope.launch {
        if (!isLoading && licenses.isEmpty()) {
            isLoading = true
            val csv = readLicensesFile()
            if (csv != null) {
                parseLicenseFile(csv)
            }
            isLoading = false
        }
    }


    /**
     * Method loads the license text for the specified license. This loads in another thread and is
     * stored to the variables "displayedLicenseName" and "displayedLicenseText".
     *
     * @param license   License whose text to load.
     */
    fun loadLicenseText(license: License) = viewModelScope.launch {
        val content: ByteArray
        try {
            val application = getApplication<Application>()
            val resources = application.resources
            val inputStream: InputStream = resources.openRawResource(resources.getIdentifier(license.licenseFile, "raw", application.packageName))
            content = ByteArray(inputStream.available())
            inputStream.read(content)
            inputStream.close()
            displayedLicenseName = license.licenseName
            displayedLicenseText = String(content)
        }
        catch (e: Exception) {
            displayedLicenseName = null
            displayedLicenseText = null
        }
    }


    /**
     * Method reads the file containing the list of all used software licenses.
     *
     * @return  Content of the file.
     */
    private fun readLicensesFile(): String? {
        val content: ByteArray
        try {
            val inputStream: InputStream = getApplication<Application>().resources.openRawResource(R.raw.licenses)
            content = ByteArray(inputStream.available())
            inputStream.read(content)
            inputStream.close()
        }
        catch (e: Exception) {
            return null
        }
        return String(content)
    }


    /**
     * Method parses the licenses file whose CSV is passed as argument. The result is written to the
     * list "licenses".
     *
     * @param csv   CSV to parse.
     */
    private fun parseLicenseFile(csv: String) {
        val list = mutableListOf<License>()
        val software: List<String> = csv.split("\n")
        software.forEach { s ->
            val attributes: List<String> = s.split(",")
            if (attributes.size == 4) {
                list.add(License(
                    softwareName = attributes[0],
                    softwareVersion = attributes[1],
                    licenseFile = attributes[2],
                    licenseName = attributes[3]
                ))
            }
        }
        licenses = list
    }

}
